// vue-files-preview/es/packages/index.mjs
import "./preview/supports/vue-files-preview/index.mjs";
import "./preview/supports/audio-preview/index.mjs";
import "./preview/supports/code-preview/index.mjs";
import "./preview/supports/docx-preview/index.mjs";
import "./preview/supports/epub-preview/index.mjs";
import "./preview/supports/md-preview/index.mjs";
import "./preview/supports/pdf-preview/index.mjs";
import "./preview/supports/pic-preview/index.mjs";
import "./preview/supports/txt-preview/index.mjs";
import "./preview/supports/video-preview/index.mjs";
import "./preview/supports/xlsx-preview/index.mjs";
import { version as e } from "./version.mjs";
import m from "./preview/supports/vue-files-preview/index.vue.mjs";
import t from "./preview/supports/audio-preview/index.vue.mjs";
import p from "./preview/supports/code-preview/index.vue.mjs";
import s from "./preview/supports/docx-preview/index.vue.mjs";
import f from "./preview/supports/epub-preview/index.vue.mjs";
import P from "./preview/supports/md-preview/index.vue.mjs";
import v from "./preview/supports/pdf-preview/index.vue.mjs";
import c from "./preview/supports/pic-preview/index.vue.mjs";
import n from "./preview/supports/txt-preview/index.vue.mjs";
import w from "./preview/supports/video-preview/index.vue.mjs";
import a from "./preview/supports/xlsx-preview/index.vue.mjs";
const r = {
  VueFilesPreview: m,
  AudioPreview: t,
  CodePreview: p,
  DocxPreview: s,
  EpubPreview: f,
  MdPreview: P,
  PdfPreview: v,
  PicPreview: c,
  TxtPreview: n,
  VideoPreview: w,
  XlsxPreview: a
};
function _(o) {
  Object.keys(r).forEach((i) => {
    o.component(i, r[i]);
  });
}
const B = { version: e, install: _ };
export {
  t as AudioPreview,
  p as CodePreview,
  s as DocxPreview,
  f as EpubPreview,
  P as MdPreview,
  v as PdfPreview,
  c as PicPreview,
  n as TxtPreview,
  w as VideoPreview,
  m as VueFilesPreview,
  a as XlsxPreview,
  B as default
};
