// highlight.js@11.10.0/lib/languages/awk.mjs
var n, r;
function b() {
  if (r) return n;
  r = 1;
  function i(e) {
    const a = {
      className: "variable",
      variants: [
        { begin: /\$[\w\d#@][\w\d_]*/ },
        { begin: /\$\{(.*?)\}/ }
      ]
    }, t = "BEGIN END if else while do for in break continue delete next nextfile function func exit|10", u = {
      className: "string",
      contains: [e.BACKSLASH_ESCAPE],
      variants: [
        {
          begin: /(u|b)?r?'''/,
          end: /'''/,
          relevance: 10
        },
        {
          begin: /(u|b)?r?"""/,
          end: /"""/,
          relevance: 10
        },
        {
          begin: /(u|r|ur)'/,
          end: /'/,
          relevance: 10
        },
        {
          begin: /(u|r|ur)"/,
          end: /"/,
          relevance: 10
        },
        {
          begin: /(b|br)'/,
          end: /'/
        },
        {
          begin: /(b|br)"/,
          end: /"/
        },
        e.APOS_STRING_MODE,
        e.QUOTE_STRING_MODE
      ]
    };
    return {
      name: "Awk",
      keywords: { keyword: t },
      contains: [
        a,
        u,
        e.REGEXP_MODE,
        e.HASH_COMMENT_MODE,
        e.NUMBER_MODE
      ]
    };
  }
  return n = i, n;
}
export {
  b as __require
};
