// highlight.js@11.10.0/lib/languages/d.mjs
var _, c;
function R() {
  if (c) return _;
  c = 1;
  function o(e) {
    const l = {
      $pattern: e.UNDERSCORE_IDENT_RE,
      keyword: "abstract alias align asm assert auto body break byte case cast catch class const continue debug default delete deprecated do else enum export extern final finally for foreach foreach_reverse|10 goto if immutable import in inout int interface invariant is lazy macro mixin module new nothrow out override package pragma private protected public pure ref return scope shared static struct super switch synchronized template this throw try typedef typeid typeof union unittest version void volatile while with __FILE__ __LINE__ __gshared|10 __thread __traits __DATE__ __EOF__ __TIME__ __TIMESTAMP__ __VENDOR__ __VERSION__",
      built_in: "bool cdouble cent cfloat char creal dchar delegate double dstring float function idouble ifloat ireal long real short string ubyte ucent uint ulong ushort wchar wstring",
      literal: "false null true"
    }, a = "(0|[1-9][\\d_]*)", n = "(0|[1-9][\\d_]*|\\d[\\d_]*|[\\d_]+?\\d)", d = "0[bB][01_]+", t = "([\\da-fA-F][\\da-fA-F_]*|_[\\da-fA-F][\\da-fA-F_]*)", E = "0[xX]" + t, r = "([eE][+-]?" + n + ")", u = "(" + n + "(\\.\\d*|" + r + ")|\\d+\\." + n + "|\\." + a + r + "?)", D = "(0[xX](" + t + "\\." + t + "|\\.?" + t + ")[pP][+-]?" + n + ")", i = "(" + a + "|" + d + "|" + E + ")", g = "(" + D + "|" + u + ")", s = `\\\\(['"\\?\\\\abfnrtv]|u[\\dA-Fa-f]{4}|[0-7]{1,3}|x[\\dA-Fa-f]{2}|U[\\dA-Fa-f]{8})|&[a-zA-Z\\d]{2,};`, N = {
      className: "number",
      begin: "\\b" + i + "(L|u|U|Lu|LU|uL|UL)?",
      relevance: 0
    }, f = {
      className: "number",
      begin: "\\b(" + g + "([fF]|L|i|[fF]i|Li)?|" + i + "(i|[fF]i|Li))",
      relevance: 0
    }, m = {
      className: "string",
      begin: "'(" + s + "|.)",
      end: "'",
      illegal: "."
    }, b = {
      className: "string",
      begin: '"',
      contains: [{
        begin: s,
        relevance: 0
      }],
      end: '"[cwd]?'
    }, O = {
      className: "string",
      begin: '[rq]"',
      end: '"[cwd]?',
      relevance: 5
    }, M = {
      className: "string",
      begin: "`",
      end: "`[cwd]?"
    }, T = {
      className: "string",
      begin: 'x"[\\da-fA-F\\s\\n\\r]*"[cwd]?',
      relevance: 10
    }, A = {
      className: "string",
      begin: 'q"\\{',
      end: '\\}"'
    }, p = {
      className: "meta",
      begin: "^#!",
      end: "$",
      relevance: 5
    }, I = {
      className: "meta",
      begin: "#(line)",
      end: "$",
      relevance: 5
    }, h = {
      className: "keyword",
      begin: "@[a-zA-Z_][a-zA-Z_\\d]*"
    }, S = e.COMMENT(
      "\\/\\+",
      "\\+\\/",
      {
        contains: ["self"],
        relevance: 10
      }
    );
    return {
      name: "D",
      keywords: l,
      contains: [
        e.C_LINE_COMMENT_MODE,
        e.C_BLOCK_COMMENT_MODE,
        S,
        T,
        b,
        O,
        M,
        A,
        f,
        N,
        m,
        p,
        I,
        h
      ]
    };
  }
  return _ = o, _;
}
export {
  R as __require
};
