// @xmldom/xmldom@0.7.13/lib/dom-parser.mjs
import { __exports as p } from "../../../../../../../_virtual/dom-parser.mjs";
import "./conventions.mjs";
import "./dom.mjs";
import "./entities.mjs";
import "./sax.mjs";
import { __exports as M } from "../../../../../../../_virtual/entities.mjs";
import { __exports as N } from "../../../../../../../_virtual/conventions.mjs";
import { __exports as y } from "../../../../../../../_virtual/dom.mjs";
import { __exports as b } from "../../../../../../../_virtual/sax.mjs";
var _ = N, h = y, v = M, g = b, T = h.DOMImplementation, E = _.NAMESPACE, S = g.ParseError, A = g.XMLReader;
function x(t) {
  this.options = t || { locator: {} };
}
x.prototype.parseFromString = function(t, n) {
  var r = this.options, e = new A(), i = r.domBuilder || new m(), a = r.errorHandler, c = r.locator, o = r.xmlns || {}, l = /\/x?html?$/.test(n), s = l ? v.HTML_ENTITIES : v.XML_ENTITIES;
  return c && i.setDocumentLocator(c), e.errorHandler = L(a, i, c), e.domBuilder = r.domBuilder || i, l && (o[""] = E.HTML), o.xml = o.xml || E.XML, t && typeof t == "string" ? e.parse(t, o, s) : e.errorHandler.error("invalid doc source"), i.doc;
};
function L(t, n, r) {
  if (!t) {
    if (n instanceof m)
      return n;
    t = n;
  }
  var e = {}, i = t instanceof Function;
  r = r || {};
  function a(c) {
    var o = t[c];
    !o && i && (o = t.length == 2 ? function(l) {
      t(c, l);
    } : t), e[c] = o && function(l) {
      o("[xmldom " + c + "]	" + l + d(r));
    } || function() {
    };
  }
  return a("warning"), a("error"), a("fatalError"), e;
}
function m() {
  this.cdata = !1;
}
function u(t, n) {
  n.lineNumber = t.lineNumber, n.columnNumber = t.columnNumber;
}
m.prototype = {
  startDocument: function() {
    this.doc = new T().createDocument(null, null, null), this.locator && (this.doc.documentURI = this.locator.systemId);
  },
  startElement: function(t, n, r, e) {
    var i = this.doc, a = i.createElementNS(t, r || n), c = e.length;
    f(this, a), this.currentElement = a, this.locator && u(this.locator, a);
    for (var o = 0; o < c; o++) {
      var t = e.getURI(o), l = e.getValue(o), r = e.getQName(o), s = i.createAttributeNS(t, r);
      this.locator && u(e.getLocator(o), s), s.value = s.nodeValue = l, a.setAttributeNode(s);
    }
  },
  endElement: function(t, n, r) {
    var e = this.currentElement;
    e.tagName, this.currentElement = e.parentNode;
  },
  startPrefixMapping: function(t, n) {
  },
  endPrefixMapping: function(t) {
  },
  processingInstruction: function(t, n) {
    var r = this.doc.createProcessingInstruction(t, n);
    this.locator && u(this.locator, r), f(this, r);
  },
  ignorableWhitespace: function(t, n, r) {
  },
  characters: function(t, n, r) {
    if (t = D.apply(this, arguments), t) {
      if (this.cdata)
        var e = this.doc.createCDATASection(t);
      else
        var e = this.doc.createTextNode(t);
      this.currentElement ? this.currentElement.appendChild(e) : /^\s*$/.test(t) && this.doc.appendChild(e), this.locator && u(this.locator, e);
    }
  },
  skippedEntity: function(t) {
  },
  endDocument: function() {
    this.doc.normalize();
  },
  setDocumentLocator: function(t) {
    (this.locator = t) && (t.lineNumber = 0);
  },
  //LexicalHandler
  comment: function(t, n, r) {
    t = D.apply(this, arguments);
    var e = this.doc.createComment(t);
    this.locator && u(this.locator, e), f(this, e);
  },
  startCDATA: function() {
    this.cdata = !0;
  },
  endCDATA: function() {
    this.cdata = !1;
  },
  startDTD: function(t, n, r) {
    var e = this.doc.implementation;
    if (e && e.createDocumentType) {
      var i = e.createDocumentType(t, n, r);
      this.locator && u(this.locator, i), f(this, i), this.doc.doctype = i;
    }
  },
  /**
   * @see org.xml.sax.ErrorHandler
   * @link http://www.saxproject.org/apidoc/org/xml/sax/ErrorHandler.html
   */
  warning: function(t) {
    console.warn("[xmldom warning]	" + t, d(this.locator));
  },
  error: function(t) {
    console.error("[xmldom error]	" + t, d(this.locator));
  },
  fatalError: function(t) {
    throw new S(t, this.locator);
  }
};
function d(t) {
  if (t)
    return `
@` + (t.systemId || "") + "#[line:" + t.lineNumber + ",col:" + t.columnNumber + "]";
}
function D(t, n, r) {
  return typeof t == "string" ? t.substr(n, r) : t.length >= n + r || n ? new java.lang.String(t, n, r) + "" : t;
}
"endDTD,startEntity,endEntity,attributeDecl,elementDecl,externalEntityDecl,internalEntityDecl,resolveEntity,getExternalSubset,notationDecl,unparsedEntityDecl".replace(/\w+/g, function(t) {
  m.prototype[t] = function() {
    return null;
  };
});
function f(t, n) {
  t.currentElement ? t.currentElement.appendChild(n) : t.doc.appendChild(n);
}
p.__DOMHandler = m;
p.DOMParser = x;
p.DOMImplementation = h.DOMImplementation;
p.XMLSerializer = h.XMLSerializer;
export {
  p as default
};
